File: Game_Controler/test/protocol_test.dart
import 'dart:typed_data';
import 'package:flutter_test/flutter_test.dart';
import 'package:game_controler/protocol.dart';

void main() {
  group('MsgTypes constants', () {
    test('should have correct values', () {
      expect(MsgTypes.move, 0x01);
      expect(MsgTypes.buttonDown, 0x02);
      expect(MsgTypes.buttonUp, 0x03);
      expect(MsgTypes.heartbeat, 0x04);
      expect(MsgTypes.batch, 0x05);
    });
  });

  group('Sides constants', () {
    test('should have correct values', () {
      expect(Sides.left, 0);
      expect(Sides.right, 1);
    });
  });

  group('Priority constants', () {
    test('should have correct values', () {
      expect(Priority.high, 2);
      expect(Priority.medium, 1);
      expect(Priority.low, 0);
    });
  });

  group('InputEvent.move', () {
    test('should create move event with correct properties', () {
      final event = InputEvent.move(
        side: 'left',
        x: 0.5,
        y: -0.3,
        timestamp: 1000,
      );

      expect(event.type, InputType.move);
      expect(event.side, 'left');
      expect(event.x, 0.5);
      expect(event.y, -0.3);
      expect(event.timestamp, 1000);
      expect(event.index, 0);
      expect(event.priority, Priority.low);
    });

    test('should have correct byteSize for move', () {
      final event = InputEvent.move(
        side: 'left',
        x: 0.5,
        y: -0.3,
        timestamp: 1000,
      );

      expect(event.byteSize, 11);
    });
  });

  group('InputEvent.button', () {
    test('should create button pressed event', () {
      final event = InputEvent.button(
        side: 'right',
        index: 0,
        pressed: true,
        timestamp: 2000,
      );

      expect(event.type, InputType.button);
      expect(event.side, 'right');
      expect(event.index, 0);
      expect(event.pressed, true);
      expect(event.timestamp, 2000);
      expect(event.priority, Priority.high);
    });

    test('should create button released event', () {
      final event = InputEvent.button(
        side: 'left',
        index: 3,
        pressed: false,
        timestamp: 3000,
      );

      expect(event.pressed, false);
      expect(event.index, 3);
    });

    test('should have correct byteSize for button', () {
      final event = InputEvent.button(
        side: 'right',
        index: 0,
        pressed: true,
        timestamp: 2000,
      );

      expect(event.byteSize, 7);
    });
  });

  group('encodeMove', () {
    test('should encode move event with correct byte length', () {
      final bytes = encodeMove('left', 0.5, -0.3, 1000);

      expect(bytes.length, 11);
    });

    test('should set correct message type', () {
      final bytes = encodeMove('left', 0.5, -0.3, 1000);

      expect(bytes[0], MsgTypes.move);
    });

    test('should set correct side for left', () {
      final bytes = encodeMove('left', 0.5, -0.3, 1000);

      expect(bytes[1], Sides.left);
    });

    test('should set correct side for right', () {
      final bytes = encodeMove('right', 0.5, -0.3, 1000);

      expect(bytes[1], Sides.right);
    });

    test('should encode x value as int16 big endian', () {
      final bytes = encodeMove('left', 1.0, 0.0, 1000);
      final expectedX = (1.0 * 32767).round();

      final data = ByteData.sublistView(bytes, 2, 4);
      expect(data.getInt16(0, Endian.big), expectedX);
    });

    test('should encode y value as int16 big endian', () {
      final bytes = encodeMove('left', 0.0, -1.0, 1000);
      final expectedY = (-1.0 * 32767).round();

      final data = ByteData.sublistView(bytes, 4, 6);
      expect(data.getInt16(0, Endian.big), expectedY);
    });

    test('should encode timestamp at correct offset', () {
      final bytes = encodeMove('left', 0.5, -0.3, 12345678);

      final data = ByteData.sublistView(bytes, 6, 10);
      expect(data.getUint32(0, Endian.big), 12345678);
    });
  });

  group('encodeButton', () {
    test('should encode button down with correct length', () {
      final bytes = encodeButton('left', 0, true, 1000);

      expect(bytes.length, 7);
    });

    test('should set buttonDown message type for pressed', () {
      final bytes = encodeButton('left', 0, true, 1000);

      expect(bytes[0], MsgTypes.buttonDown);
    });

    test('should set buttonUp message type for released', () {
      final bytes = encodeButton('left', 0, false, 1000);

      expect(bytes[0], MsgTypes.buttonUp);
    });

    test('should set correct side byte', () {
      final bytes = encodeButton('right', 5, true, 1000);

      expect(bytes[1], Sides.right);
    });

    test('should set button index at correct offset', () {
      final bytes = encodeButton('left', 7, true, 1000);

      expect(bytes[2], 7);
    });

    test('should encode timestamp at correct offset', () {
      final bytes = encodeButton('left', 0, true, 9876543);

      final data = ByteData.sublistView(bytes, 3, 7);
      expect(data.getUint32(0, Endian.big), 9876543);
    });
  });

  group('encodeHeartbeat', () {
    test('should encode heartbeat with correct length', () {
      final bytes = encodeHeartbeat(1000);

      expect(bytes.length, 5);
    });

    test('should set heartbeat message type', () {
      final bytes = encodeHeartbeat(1000);

      expect(bytes[0], MsgTypes.heartbeat);
    });

    test('should encode timestamp at correct offset', () {
      final bytes = encodeHeartbeat(5555555);

      final data = ByteData.sublistView(bytes, 1, 5);
      expect(data.getUint32(0, Endian.big), 5555555);
    });
  });

  group('encodeBatch', () {
    test('should encode empty batch as empty Uint8List', () {
      final bytes = encodeBatch([]);

      expect(bytes.length, 0);
    });

    test('should set batch message type', () {
      final bytes = encodeBatch([
        InputEvent.move(side: 'left', x: 0.5, y: 0.0, timestamp: 1000),
      ]);

      expect(bytes[0], MsgTypes.batch);
    });

    test('should set correct event count', () {
      final bytes = encodeBatch([
        InputEvent.move(side: 'left', x: 0.5, y: 0.0, timestamp: 1000),
        InputEvent.button(
            side: 'right', index: 0, pressed: true, timestamp: 1001),
      ]);

      expect(bytes[1], 2);
    });

    test('should encode multiple events correctly', () {
      final events = [
        InputEvent.move(side: 'left', x: 0.5, y: 0.0, timestamp: 1000),
        InputEvent.button(
            side: 'right', index: 0, pressed: true, timestamp: 1001),
      ];
      final bytes = encodeBatch(events);

      final expectedLength = 2 + 11 + 7;
      expect(bytes.length, expectedLength);
    });

    test('should preserve event order in batch', () {
      final events = [
        InputEvent.move(side: 'left', x: 0.1, y: 0.2, timestamp: 1000),
        InputEvent.move(side: 'right', x: 0.3, y: 0.4, timestamp: 1001),
      ];
      final bytes = encodeBatch(events);

      expect(bytes[2], MsgTypes.move);
      expect(bytes[13], MsgTypes.move);
    });
  });

  group('InputEvent.toBytes', () {
    test('should convert move event to bytes', () {
      final event = InputEvent.move(
        side: 'left',
        x: 0.5,
        y: -0.3,
        timestamp: 1000,
      );

      final bytes = event.toBytes();

      expect(bytes.length, 11);
      expect(bytes[0], MsgTypes.move);
      expect(bytes[1], Sides.left);
    });

    test('should convert pressed button event to bytes', () {
      final event = InputEvent.button(
        side: 'right',
        index: 5,
        pressed: true,
        timestamp: 2000,
      );

      final bytes = event.toBytes();

      expect(bytes.length, 7);
      expect(bytes[0], MsgTypes.buttonDown);
      expect(bytes[2], 5);
    });

    test('should convert released button event to bytes', () {
      final event = InputEvent.button(
        side: 'left',
        index: 3,
        pressed: false,
        timestamp: 3000,
      );

      final bytes = event.toBytes();

      expect(bytes[0], MsgTypes.buttonUp);
    });
  });

  group('round trip encoding', () {
    test('should encode and decode move values correctly', () {
      final originalX = 0.5;
      final originalY = -0.75;

      final bytes = encodeMove('left', originalX, originalY, 1000);

      final data = ByteData.sublistView(bytes, 2, 6);
      final decodedX = data.getInt16(0, Endian.big) / 32767;
      final decodedY = data.getInt16(2, Endian.big) / 32767;

      expect(decodedX, closeTo(originalX, 0.001));
      expect(decodedY, closeTo(originalY, 0.001));
    });

    test('should encode max values correctly', () {
      final bytes = encodeMove('left', 1.0, 1.0, 1000);

      final data = ByteData.sublistView(bytes, 2, 6);
      expect(data.getInt16(0, Endian.big), 32767);
      expect(data.getInt16(2, Endian.big), 32767);
    });

    test('should encode min values correctly', () {
      final bytes = encodeMove('left', -1.0, -1.0, 1000);

      final data = ByteData.sublistView(bytes, 2, 6);
      expect(data.getInt16(0, Endian.big), -32767);
      expect(data.getInt16(2, Endian.big), -32767);
    });

    test('should encode zero values correctly', () {
      final bytes = encodeMove('left', 0.0, 0.0, 1000);

      final data = ByteData.sublistView(bytes, 2, 6);
      expect(data.getInt16(0, Endian.big), 0);
      expect(data.getInt16(2, Endian.big), 0);
    });
  });
}

File: Backend_controler/test/protocol.test.js
const assert = require('assert');
const { MSG_TYPES, SIDES, decodeMessage, isBinaryMessage } = require('../protocol');

describe('Protocol Constants', () => {
  describe('MSG_TYPES', () => {
    it('should have correct MOVE value', () => {
      assert.strictEqual(MSG_TYPES.MOVE, 0x01);
    });

    it('should have correct BUTTON_DOWN value', () => {
      assert.strictEqual(MSG_TYPES.BUTTON_DOWN, 0x02);
    });

    it('should have correct BUTTON_UP value', () => {
      assert.strictEqual(MSG_TYPES.BUTTON_UP, 0x03);
    });

    it('should have correct HEARTBEAT value', () => {
      assert.strictEqual(MSG_TYPES.HEARTBEAT, 0x04);
    });

    it('should have correct BATCH value', () => {
      assert.strictEqual(MSG_TYPES.BATCH, 0x05);
    });
  });

  describe('SIDES', () => {
    it('should have correct left value', () => {
      assert.strictEqual(SIDES.left, 0);
    });

    it('should have correct right value', () => {
      assert.strictEqual(SIDES.right, 1);
    });
  });
});

describe('decodeMessage', () => {
  describe('MOVE message', () => {
    it('should decode MOVE message correctly', () => {
      const buffer = Buffer.alloc(11);
      buffer[0] = MSG_TYPES.MOVE;
      buffer[1] = SIDES.left;
      buffer.writeInt16BE(10000, 2);
      buffer.writeInt16BE(-5000, 4);
      buffer.writeUInt32BE(12345678, 6);

      const result = decodeMessage(buffer);

      assert.strictEqual(result.type, 'move');
      assert.strictEqual(result.side, 'left');
      assert.ok(Math.abs(result.x - 10000 / 32767) < 0.01);
      assert.ok(Math.abs(result.y - (-5000 / 32767)) < 0.01);
      assert.strictEqual(result.timestamp, 12345678);
    });

    it('should decode right side', () => {
      const buffer = Buffer.alloc(11);
      buffer[0] = MSG_TYPES.MOVE;
      buffer[1] = SIDES.right;
      buffer.writeInt16BE(32767, 2);
      buffer.writeInt16BE(32767, 4);
      buffer.writeUInt32BE(1000, 6);

      const result = decodeMessage(buffer);

      assert.strictEqual(result.side, 'right');
      assert.strictEqual(result.x, 1);
      assert.strictEqual(result.y, 1);
    });

    it('should return null for incomplete MOVE message', () => {
      const buffer = Buffer.alloc(5);
      buffer[0] = MSG_TYPES.MOVE;

      const result = decodeMessage(buffer);

      assert.strictEqual(result, null);
    });

    it('should handle negative values', () => {
      const buffer = Buffer.alloc(11);
      buffer[0] = MSG_TYPES.MOVE;
      buffer[1] = SIDES.left;
      buffer.writeInt16BE(-32767, 2);
      buffer.writeInt16BE(-32767, 4);
      buffer.writeUInt32BE(1000, 6);

      const result = decodeMessage(buffer);

      assert.strictEqual(result.x, -1);
      assert.strictEqual(result.y, -1);
    });
  });

  describe('BUTTON_DOWN message', () => {
    it('should decode BUTTON_DOWN message correctly', () => {
      const buffer = Buffer.alloc(7);
      buffer[0] = MSG_TYPES.BUTTON_DOWN;
      buffer[1] = SIDES.left;
      buffer[2] = 0;
      buffer.writeUInt32BE(9876543, 3);

      const result = decodeMessage(buffer);

      assert.strictEqual(result.type, 'button_down');
      assert.strictEqual(result.side, 'left');
      assert.strictEqual(result.index, 0);
      assert.strictEqual(result.timestamp, 9876543);
    });

    it('should decode right side button', () => {
      const buffer = Buffer.alloc(7);
      buffer[0] = MSG_TYPES.BUTTON_DOWN;
      buffer[1] = SIDES.right;
      buffer[2] = 5;
      buffer.writeUInt32BE(1000, 3);

      const result = decodeMessage(buffer);

      assert.strictEqual(result.side, 'right');
      assert.strictEqual(result.index, 5);
    });

    it('should return null for incomplete BUTTON_DOWN', () => {
      const buffer = Buffer.alloc(3);
      buffer[0] = MSG_TYPES.BUTTON_DOWN;

      const result = decodeMessage(buffer);

      assert.strictEqual(result, null);
    });
  });

  describe('BUTTON_UP message', () => {
    it('should decode BUTTON_UP message correctly', () => {
      const buffer = Buffer.alloc(7);
      buffer[0] = MSG_TYPES.BUTTON_UP;
      buffer[1] = SIDES.left;
      buffer[2] = 3;
      buffer.writeUInt32BE(5555555, 3);

      const result = decodeMessage(buffer);

      assert.strictEqual(result.type, 'button_up');
      assert.strictEqual(result.side, 'left');
      assert.strictEqual(result.index, 3);
      assert.strictEqual(result.timestamp, 5555555);
    });
  });

  describe('HEARTBEAT message', () => {
    it('should decode HEARTBEAT message correctly', () => {
      const buffer = Buffer.alloc(5);
      buffer[0] = MSG_TYPES.HEARTBEAT;
      buffer.writeUInt32BE(1111111, 1);

      const result = decodeMessage(buffer);

      assert.strictEqual(result.type, 'heartbeat');
      assert.strictEqual(result.timestamp, 1111111);
    });

    it('should return null for incomplete HEARTBEAT', () => {
      const buffer = Buffer.alloc(1);
      buffer[0] = MSG_TYPES.HEARTBEAT;

      const result = decodeMessage(buffer);

      assert.strictEqual(result, null);
    });
  });

  describe('BATCH message', () => {
    it('should decode empty BATCH message', () => {
      const buffer = Buffer.alloc(3);
      buffer[0] = MSG_TYPES.BATCH;
      buffer[1] = 0;
      buffer[2] = 0;

      const result = decodeMessage(buffer);

      assert.strictEqual(result === null || result?.type === 'batch', true);
    });

    it('should decode BATCH with single MOVE event', () => {
      const buffer = Buffer.alloc(13);
      buffer[0] = MSG_TYPES.BATCH;
      buffer[1] = 1;
      buffer[2] = MSG_TYPES.MOVE;
      buffer[3] = SIDES.left;
      buffer.writeInt16BE(10000, 4);
      buffer.writeInt16BE(20000, 6);
      buffer.writeUInt32BE(1000, 8);

      const result = decodeMessage(buffer);

      assert.strictEqual(result.type, 'batch');
      assert.strictEqual(result.events.length, 1);
      assert.strictEqual(result.events[0].type, 'move');
    });

    it('should decode BATCH with multiple events', () => {
      const buffer = Buffer.alloc(20);
      buffer[0] = MSG_TYPES.BATCH;
      buffer[1] = 2;

      buffer[2] = MSG_TYPES.MOVE;
      buffer[3] = SIDES.left;
      buffer.writeInt16BE(10000, 4);
      buffer.writeInt16BE(20000, 6);
      buffer.writeUInt32BE(1000, 8);

      buffer[12] = MSG_TYPES.BUTTON_DOWN;
      buffer[13] = SIDES.right;
      buffer[14] = 0;
      buffer.writeUInt32BE(1001, 15);

      const result = decodeMessage(buffer);

      assert.ok(result !== null);
      assert.strictEqual(result.events[0].type, 'move');
      if (result.events[1]) {
        assert.strictEqual(result.events[1].type, 'button_down');
      }
    });
  });

  describe('edge cases', () => {
    it('should return null for null buffer', () => {
      const result = decodeMessage(null);
      assert.strictEqual(result, null);
    });

    it('should return null for empty buffer', () => {
      const result = decodeMessage(Buffer.alloc(0));
      assert.strictEqual(result, null);
    });

    it('should return null for buffer too short', () => {
      const buffer = Buffer.alloc(3);
      buffer[0] = MSG_TYPES.MOVE;

      const result = decodeMessage(buffer);

      assert.strictEqual(result, null);
    });

    it('should return null for unknown message type', () => {
      const buffer = Buffer.alloc(5);
      buffer[0] = 0xFF;
      buffer.writeUInt32BE(1000, 1);

      const result = decodeMessage(buffer);

      assert.strictEqual(result, null);
    });

    it('should default to left side for unknown side index', () => {
      const buffer = Buffer.alloc(11);
      buffer[0] = MSG_TYPES.MOVE;
      buffer[1] = 99;
      buffer.writeInt16BE(1000, 2);
      buffer.writeInt16BE(1000, 4);
      buffer.writeUInt32BE(1000, 6);

      const result = decodeMessage(buffer);

      assert.strictEqual(result.side, 'left');
    });
  });
});

describe('isBinaryMessage', () => {
  it('should return true for MOVE', () => {
    const buffer = Buffer.from([MSG_TYPES.MOVE, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]);
    assert.strictEqual(isBinaryMessage(buffer), true);
  });

  it('should return true for BUTTON_DOWN', () => {
    const buffer = Buffer.from([MSG_TYPES.BUTTON_DOWN, 0, 0, 0, 0, 0, 0, 0]);
    assert.strictEqual(isBinaryMessage(buffer), true);
  });

  it('should return true for BUTTON_UP', () => {
    const buffer = Buffer.from([MSG_TYPES.BUTTON_UP, 0, 0, 0, 0, 0, 0, 0]);
    assert.strictEqual(isBinaryMessage(buffer), true);
  });

  it('should return true for HEARTBEAT', () => {
    const buffer = Buffer.from([MSG_TYPES.HEARTBEAT, 0, 0, 0, 0]);
    assert.strictEqual(isBinaryMessage(buffer), true);
  });

  it('should return true for BATCH', () => {
    const buffer = Buffer.from([MSG_TYPES.BATCH, 0, 0]);
    assert.strictEqual(isBinaryMessage(buffer), true);
  });

  it('should return false for null buffer', () => {
    assert.strictEqual(isBinaryMessage(null), false);
  });

  it('should return false for empty buffer', () => {
    assert.strictEqual(isBinaryMessage(Buffer.alloc(0)), false);
  });

  it('should return false for non-binary message type', () => {
    const buffer = Buffer.from([0xFF, 0, 0, 0, 0]);
    assert.strictEqual(isBinaryMessage(buffer), false);
  });
});

describe('round trip encoding/decoding', () => {
  it('should encode and decode MOVE with exact values', () => {
    const originalX = 0.5;
    const originalY = -0.75;
    const timestamp = 12345678;

    const buffer = Buffer.alloc(11);
    buffer[0] = MSG_TYPES.MOVE;
    buffer[1] = SIDES.left;
    buffer.writeInt16BE(Math.round(originalX * 32767), 2);
    buffer.writeInt16BE(Math.round(originalY * 32767), 4);
    buffer.writeUInt32BE(timestamp, 6);

    const result = decodeMessage(buffer);

    assert.ok(Math.abs(result.x - originalX) < 0.01);
    assert.ok(Math.abs(result.y - originalY) < 0.01);
    assert.strictEqual(result.timestamp, timestamp);
  });

  it('should handle full range values', () => {
    const maxBuffer = Buffer.alloc(11);
    maxBuffer[0] = MSG_TYPES.MOVE;
    maxBuffer[1] = SIDES.left;
    maxBuffer.writeInt16BE(32767, 2);
    maxBuffer.writeInt16BE(32767, 4);
    maxBuffer.writeUInt32BE(1000, 6);

    const maxResult = decodeMessage(maxBuffer);
    assert.strictEqual(maxResult.x, 1);
    assert.strictEqual(maxResult.y, 1);

    const minBuffer = Buffer.alloc(11);
    minBuffer[0] = MSG_TYPES.MOVE;
    minBuffer[1] = SIDES.left;
    minBuffer.writeInt16BE(-32767, 2);
    minBuffer.writeInt16BE(-32767, 4);
    minBuffer.writeUInt32BE(1000, 6);

    const minResult = decodeMessage(minBuffer);
    assert.strictEqual(minResult.x, -1);
    assert.strictEqual(minResult.y, -1);
  });
});

